toolchain/check/testdata/impl/no_prelude/name_poisoning.carbon

```diff
@@ -0,0 +1,142 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/impl/no_prelude/name_poisoning.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/impl/no_prelude/name_poisoning.carbon
+
+// --- fail_impl_function_poisoned.carbon
+
+library "[[@TEST_NAME]]";
+
+interface I {
+  fn F();
+}
+
+class F {}
+
+class C1 {
+  impl as I {
+    class C2 {
+      // CHECK:STDERR: fail_impl_function_poisoned.carbon:[[@LINE+3]]:14: error: name used before it was declared [NameUseBeforeDecl]
+      // CHECK:STDERR:       var v: F;
+      // CHECK:STDERR:              ^
+      var v: F;
+    }
```

Comment:
It's probably best to use an example here that would otherwise be valid if possible, and it's not valid to declare things in the `impl` that don't correspond to names in the interface. How about something like this:

```carbon
interface I {
  fn A(x: Self);
  fn B();
}

class B {
  impl as I {
    fn A(x: B);
    fn B();
  }
}
```

Longer term, I think the plan is for name lookup for the impl scope to look in the facet (`B as I` in this case), so it would find the function name `B` from the interface here instead of the class name. It's invalid either way, though, and seems good to test.

Reply:
Done.
Interestingly, because B is poisoned, we also get `ImplMissingFunction` error.